lib/rspec/rails/matchers/have_reported_error.rb

```diff
@@ -0,0 +1,271 @@
+module RSpec
+  module Rails
+    module Matchers
+      # @api private
+      # Sentinel value to distinguish between no argument passed vs explicitly passed nil.
+      # This follows the same pattern as RSpec's raise_error matcher.
+      UndefinedValue = Object.new.freeze
+
+      # @api private
+      class ErrorSubscriber
+        attr_reader :events
+
+        ErrorEvent = Struct.new(:error, :attributes)
+
+        def initialize
+          @events = []
+        end
+
+        def report(error, **attrs)
+          @events << ErrorEvent.new(error, attrs.with_indifferent_access)
+        end
+      end
+
+      # Matcher class for `have_reported_error`. Should not be instantiated directly.
+      #
+      # Provides a way to test that an error was reported to Rails.error.
+      #
+      # @api private
+      # @see RSpec::Rails::Matchers#have_reported_error
+      class HaveReportedError < RSpec::Rails::Matchers::BaseMatcher
+        # Uses UndefinedValue as default to distinguish between no argument
+        # passed vs explicitly passed nil.
+        #
+        # @param expected_error_or_message [Class, String, Regexp, nil]
+        #   Error class, message string, or message pattern
+        # @param expected_message [String, Regexp, nil]
+        #   Expected message when first param is a class
+        def initialize(expected_error_or_message = UndefinedValue, expected_message = nil)
+          @attributes = {}
+          @warn_about_nil_error = expected_error_or_message.nil?
+
+          case expected_error_or_message
+          when UndefinedValue
+            @expected_error = nil
+            @expected_message = expected_message
+          when String, Regexp
+            @expected_error = nil
+            @expected_message = expected_error_or_message
+          else
+            @expected_error = expected_error_or_message
+            @expected_message = expected_message
+          end
+        end
+
+        def with_context(expected_attributes)
+          @attributes.merge!(expected_attributes)
+          self
+        end
+
+        def and(_)
+          raise ArgumentError, "Chaining is not supported"
+        end
+
+        def matches?(block)
+          if block.nil?
+            raise ArgumentError, "this matcher doesn't work with value expectations"
```

Comment:
Does this check have the same effect as defining a method `supports_value_expectations?` that returns false? It should be.

Reply:
RSpec is actually allowing the value expectation to reach the `matches?` method, even with `supports_value_expectations?` defined as false. This mechanism is NOT preventing the value expectation from reaching `matches?`. The explicit check was necessary in my testing! But maybe I missed something?

Reply:
This can be removed now and any test marked as pending with the reason instead.

Reply:
Done.